### ISSUE TITLE
Promote rules, once again

### ISSUE THREAD

**Issue by HechtiDerLachs** [oscar-system/Oscar.jl#2129]
I am currently struggling with the following problem:
```julia
R, (x,y) = QQ["x", "y"]
I = ideal(R, x)
Q, _ = quo(R, I)
X = Spec(Q)
U = SpecOpen(X, [y])
W = OO(U)

P, (u,v) = W["u", "v"]
PQ, _ = quo(P, ideal(P, [u]))

evaluate(v, gens(PQ))
```
For me this throws an error with the following stacktrace:
```
ERROR: LoadError: MethodError: no method matching (::QQMPolyRing)(::MPolyQuoRingElem{AbstractAlgebra.Generic.MPoly{SpecOpenRingElem{SpecOpenRing{Spec{QQField, MPolyQuoRing{QQMPolyRingElem}}, SpecOpen{Spec{QQField, MPolyQuoRing{QQMPolyRingElem}}, QQField}}}}})
Closest candidates are:
  (::MPolyRing)(::RingElem, ::RingElem; check) at ~/Oscar.jl/src/AlgebraicGeometry/Schemes/SpecOpen/Rings/Methods.jl:185
  (::QQMPolyRing)() at ~/.julia/packages/Nemo/u0XpC/src/flint/fmpq_mpoly.jl:1170
  (::AbstractAlgebra.Ring)(::Singular.n_RingElem{Singular.RingElemWrapper{S, T}}) where {S, T} at ~/.julia/packages/Singular/T7HQA/src/number/n_unknown.jl:379
  ...
Stacktrace:
  [1] (::MPolyQuoLocRing{QQField, QQFieldElem, QQMPolyRing, QQMPolyRingElem, MPolyPowersOfElement{QQField, QQFieldElem, QQMPolyRing, QQMPolyRingElem}})(f::MPolyQuoRingElem{AbstractAlgebra.Generic.MPoly{SpecOpenRingElem{SpecOpenRing{Spec{QQField, MPolyQuoRing{QQMPolyRingElem}}, SpecOpen{Spec{QQField, MPolyQuoRing{QQMPolyRingElem}}, QQField}}}}})
    @ Oscar ~/Oscar.jl/src/Rings/localization_interface.jl:421
  [2] (::Oscar.var"#3049#3050"{MPolyQuoRingElem{AbstractAlgebra.Generic.MPoly{SpecOpenRingElem{SpecOpenRing{Spec{QQField, MPolyQuoRing{QQMPolyRingElem}}, SpecOpen{Spec{QQField, MPolyQuoRing{QQMPolyRingElem}}, QQField}}}}}})(U::PrincipalOpenSubset{QQField, MPolyQuoLocRing{QQField, QQFieldElem, QQMPolyRing, QQMPolyRingElem, MPolyPowersOfElement{QQField, QQFieldElem, QQMPolyRing, QQMPolyRingElem}}, Spec{QQField, MPolyQuoRing{QQMPolyRingElem}}})
    @ Oscar ./none:0
  [3] iterate
    @ ./generator.jl:47 [inlined]
  [4] collect(itr::Base.Generator{Vector{AbsSpec}, Oscar.var"#3049#3050"{MPolyQuoRingElem{AbstractAlgebra.Generic.MPoly{SpecOpenRingElem{SpecOpenRing{Spec{QQField, MPolyQuoRing{QQMPolyRingElem}}, SpecOpen{Spec{QQField, MPolyQuoRing{QQMPolyRingElem}}, QQField}}}}}}})
    @ Base ./array.jl:787
  [5] (::SpecOpenRing{Spec{QQField, MPolyQuoRing{QQMPolyRingElem}}, SpecOpen{Spec{QQField, MPolyQuoRing{QQMPolyRingElem}}, QQField}})(f::MPolyQuoRingElem{AbstractAlgebra.Generic.MPoly{SpecOpenRingElem{SpecOpenRing{Spec{QQField, MPolyQuoRing{QQMPolyRingElem}}, SpecOpen{Spec{QQField, MPolyQuoRing{QQMPolyRingElem}}, QQField}}}}})
    @ Oscar ~/Oscar.jl/src/AlgebraicGeometry/Schemes/SpecOpen/Rings/Constructors.jl:21
  [6] try_promote
    @ ~/.julia/packages/AbstractAlgebra/YQjGv/src/NCRings.jl:43 [inlined]
  [7] promote
    @ ~/.julia/packages/AbstractAlgebra/YQjGv/src/NCRings.jl:52 [inlined]
  [8] *
    @ ~/.julia/packages/AbstractAlgebra/YQjGv/src/NCRings.jl:78 [inlined]
  [9] evaluate(a::AbstractAlgebra.Generic.MPoly{SpecOpenRingElem{SpecOpenRing{Spec{QQField, MPolyQuoRing{QQMPolyRingElem}}, SpecOpen{Spec{QQField, MPolyQuoRing{QQMPolyRingElem}}, QQField}}}}, vals::Vector{MPolyQuoRingElem{AbstractAlgebra.Generic.MPoly{SpecOpenRingElem{SpecOpenRing{Spec{QQField, MPolyQuoRing{QQMPolyRingElem}}, SpecOpen{Spec{QQField, MPolyQuoRing{QQMPolyRingElem}}, QQField}}}}}})
    @ AbstractAlgebra ~/.julia/packages/AbstractAlgebra/YQjGv/src/MPoly.jl:808
 [10] top-level scope
    @ /tmp/issue.jl:11
 [11] include(fname::String)
    @ Base.MainInclude ./client.jl:476
 [12] top-level scope
    @ REPL[3]:1
in expression starting at /tmp/issue.jl:11
```
It seems to me that the problem occurs in [9] when julia finds there's no prescribed method for multiplying coefficients of `a` (which are of type `SpecOpenRingElem`) with elements in `vals` (of type `MPolyQuoRingElem{AbstractAlgebra.Generic.MPoly{SpecOpenRingElem}}`). Apparently, the coercion of `SpecOpenRingElem` is not visible through the two layers. Am I doing something wrong, or is some `promote_rule` missing again? 

I am aware of the problematic promote rule in `src/AlgebraicGeometry/Schemes/SpecOpen/Rings/Methods.jl`, line 163, which gives precedence to `SpecOpenRingElem` over everything else. But in the above case, I was hoping for the polynomial ring to win over that. Why doesn't it?

**Comment by thofma** [comment]
Yes, your promote rules in `src/AlgebraicGeometry/Schemes/SpecOpen/Rings/Methods.jl` are very wrong. You are basically saying, that, `SpecOpenRingElem` multiplied with anything will always be a `SpecOpenRingElem`, which is not true. Of course, deleting this method solves the issue here.

This is really a bug in the implementation of `SpecOpenRing`. It is unrelated to quotients or evaluation. A simpler way to reproduce it:
```
R, (x,y) = QQ["x", "y"]
I = ideal(R, x)
Q, _ = quo(R, I)
X = Spec(Q)
U = SpecOpen(X, [y])
W = OO(U)

P, (u,v) = W["u", "v"]
PP, (uu, vv) = P["uu", "vv"]
one(W) * uu
[boom]
```

I am inclined to close this issue. What do you think?

P.S.: The promotion system is not working by two objects competing and one winning. It will just call `promote_rule(::SpecOpenRingElem, :MPolyQuoRingElem)` (I am omitting the type parameters for our sanity), which always gives `SpecOpenRingElem`. So it will try to promote everything to a `SpecOpenRingElem`, i.e., an element of `W`. The rule you added is unsound and makes everything (which comes from `SpecOpenRingElem`) inconsistent.

**Comment by HechtiDerLachs** [comment]
Thanks for having a look. It's good to have explained that I really did something wrong there. I guess, I already had a feeling for it, but eventually I just wanted to get things running back then and considered this a "legitimate hack" or something. I will clean this up, make a pr and then resolve this issue.

**Comment by thofma** [comment]
Fixed in #2140.
